### PR TITLE
Add more HTTP tests

### PR DESCRIPTION
**Pull request by davidsh**
Ported some missing tests to GitHub. Cleaned up some of the existing tests.

### CODE REVIEW

**Comment by davidsh** [comment]
@stephentoub @pgavlin @CIPop PTAL

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs [hunk 1]
```diff
@@ -31,11 +32,33 @@ public void Dispose()
...
+            Assert.Equal(WindowsProxyUsePolicy.UseWinHttpProxy, handler.WindowsProxyUsePolicy);
+            Assert.Equal(CredentialCache.DefaultCredentials, handler.DefaultProxyCredentials);
+            Assert.Equal(null, handler.Proxy);
+            Assert.Equal(Int32.MaxValue, handler.MaxConnectionsPerServer);
+            Assert.Equal(TimeSpan.FromSeconds(60), handler.ConnectTimeout);
+            Assert.Equal(TimeSpan.FromSeconds(30), handler.SendTimeout);
+            Assert.Equal(TimeSpan.FromSeconds(30), handler.ReceiveHeadersTimeout);
+            Assert.Equal(TimeSpan.FromSeconds(30), handler.ReceiveDataTimeout);
+            Assert.Equal(64 * 1024, handler.MaxResponseHeadersLength);
+            Assert.Equal(64 * 1024, handler.MaxResponseDrainSize);
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs]: Is there a version of this in the HttpClient tests that verifies such defaults for the properties exposed from HttpClientHandler?

**Comment by stephentoub** [comment]
LGTM

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs [hunk 2]
```diff
@@ -31,11 +32,33 @@ public void Dispose()
...
+            Assert.Equal(false, handler.CheckCertificateRevocationList);
+            Assert.Equal(ClientCertificateOption.Manual, handler.ClientCertificateOption);
+            X509Certificate2Collection certs = handler.ClientCertificates;
+            Assert.True(certs.Count == 0);
+            Assert.Equal(false, handler.PreAuthenticate);
+            Assert.Equal(null, handler.ServerCredentials);
+            Assert.Equal(WindowsProxyUsePolicy.UseWinHttpProxy, handler.WindowsProxyUsePolicy);
+            Assert.Equal(CredentialCache.DefaultCredentials, handler.DefaultProxyCredentials);
+            Assert.Equal(null, handler.Proxy);
+            Assert.Equal(Int32.MaxValue, handler.MaxConnectionsPerServer);
```
Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs]: Do we want to have MaxConnections even if the RFC specifies something else? (Is this the WinHTTP default?)

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs (hunk 1 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs]: >Is there a version of this in the HttpClient tests that verifies such defaults for the properties exposed from HttpClientHandler?

Yes but not yet in GitHub.  That is next on my list for the next PR. I'm working thru consolidating the tests and beefing up the test collateral on GitHub in preparation for #3000.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs [hunk 3]
```diff
@@ -32,6 +32,10 @@ public class HttpClientHandlerTest
 
         private readonly NetworkCredential _credential = new NetworkCredential(Username, Password);
 
+        public readonly static object[][] GetServers = HttpTestServers.GetServers;
```
Comment (CIPop) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Why are these public?

**Comment by CIPop** [comment]
Two clarifying questions.

LGTM otherwise.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 3 above)
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Xunit demands that they be public. Must have something to do with how they use reflection.

This is what happens when they are not public:

`
    System.Net.Http.Functional.Tests.HttpClientHandlerTest.PostAsync_CallMethod_NullContent [FAIL]
        System.ArgumentException : Could not find public static member (property, field, or method) named 'PostServers' on System.Net.Http.Functional.Tests.HttpClientHandlerTest
`

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs (hunk 2 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs]: We allow developers to adjust this property based on what they want.  This test simply validates the current WinHttpHandler code.

If you are suggesting that we either shouldn't have the `MaxConnectionsPerServer` property at all, or perhaps have a different default value, then we should open up a separate issue to discuss.

In terms of your question:
>Is this the WinHTTP default?

The WinHTTP default is something like "2" I think. So, WinHttpHandler is adjusting that to this value as the "default" for the handler.

Comment (CIPop) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs]: Opened #3729 to discuss the default: my suggestion is to get a number maybe >2 but definitely not allow an infinite number of connections by default.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs (hunk 3 above)
Comment (CIPop) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: Thanks for explaining! I'm not very familiar with XUnit's Theory.

Comment (pgavlin) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs]: > Xunit demands that they be public. Must have something to do with how they use reflection.

Reflecting over non-public members is a strongly discouraged practice. IIRC such members many not even be available in certain conditions.

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs (hunk 2 above)
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs]: @CIPop 
>but definitely not allow an infinite number of connections by default.

`Int32.MaxValue` is actually NOT infinite.  But it is a big number. Infinite would actually be UInt32.MaxValue or (-1).  

WinHttp treats UInt32.MaxValue very special.  It allows for a lot of connections and also disable session separation for those connection pools. Our object model demanded that we separate the connection pools to be per-HttpClient/per-Handler.  So, we have to always use (UInt32.MaxValue - 1).

The property itself is signed and so it was easier to just cap it at Int32.MaxValue for ArgumentOutOfRangeChecks *and* to also make sure we keep the semantics of WinHTTP separating out the connection pools.

**Pull request merged**